### PR TITLE
Added O(1) LCA implementation

### PR DESCRIPTION
**Pull request by eysbutno**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: content/4_Gold/Tree_Euler.mdx [hunk 1]
```diff
@@ -715,6 +715,71 @@ preprocessing and query in $\mathcal{O}(1)$ time.
 
 <FocusProblem problem="rmqSample" />
 
+The following is an example implementation of a sparse table and code that
+answers LCA queries. Build time is $\mathcal{O}(N\log N)$, and queries are
+$\mathcal{O}(1)$.
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:720]: honestly i'm not sure if this is really needed

Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:720]: hmm, true. i just wrote one for fun and decided to PR it, so i can close if we don't need it

Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx:720]: wait what am i saying this literally the solution to a focus problem my bad g

File: content/4_Gold/Tree_Euler.mdx [hunk 2]
```diff
@@ -715,6 +715,77 @@ preprocessing and query in $\mathcal{O}(1)$ time.
...
+		int i = lvl(r - l + 1);
+		return min(st[i][l], st[i][r - (1 << i) + 1]);
+	}
+};
+
+class LCA {
+  private:
+	const int n;
+	const vector<vector<int>> &adj;
+	RMQ<array<int, 2>> rmq;
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:755]: in this context i think pair is both more expected and readable

File: content/4_Gold/Tree_Euler.mdx [hunk 3]
```diff
@@ -715,6 +715,77 @@ preprocessing and query in $\mathcal{O}(1)$ time.
...
+		st[0] = v;
+		for (int i = 1; i < log2dist; i++) {
+			st[i].resize(n - (1 << i) + 1);
+			for (int j = 0; j + (1 << i) <= n; j++) {
+				st[i][j] = min(st[i - 1][j], st[i - 1][j + (1 << (i - 1))]);
+			}
+		}
+	}
+
+	T query(int l, int r) {  // RMQ on [l, r]
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:745]: doc comemnts would be nice

File: content/4_Gold/Tree_Euler.mdx [hunk 4]
```diff
@@ -715,6 +715,77 @@ preprocessing and query in $\mathcal{O}(1)$ time.
...
+<CPPSection>
+
+```cpp
+template <class T> struct RMQ {
+	int n, log2dist;
+	vector<vector<T>> st;
+
+	int lvl(int x) { return 31 - __builtin_clz(x); }
+
+	void init(vector<T> &v) {
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:732]: constructors exist

File: content/4_Gold/Tree_Euler.mdx [hunk 5]
```diff
@@ -715,6 +715,77 @@ preprocessing and query in $\mathcal{O}(1)$ time.
...
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+template <class T> struct RMQ {
+	int n, log2dist;
+	vector<vector<T>> st;
+
+	int lvl(int x) { return 31 - __builtin_clz(x); }
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:730]: uwu what's this

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Tree_Euler.mdx (hunk 2 above)
Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx]: true, changed it

File: content/4_Gold/Tree_Euler.mdx (hunk 5 above)
Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx]: skull

File: content/4_Gold/Tree_Euler.mdx [hunk 6]
```diff
@@ -715,6 +715,81 @@ preprocessing and query in $\mathcal{O}(1)$ time.
...
 
+The following is an example implementation of a sparse table and code that
+answers LCA queries. Build time is $\mathcal{O}(N\log N)$, and queries are
+$\mathcal{O}(1)$.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+template <class T> struct RMQ {
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:726]: huh weird what''s w/ `class` instead of `typename`
i'm just curious bc i always use the latter

Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:726]: oh btw you forgot the headers

File: content/4_Gold/Tree_Euler.mdx [hunk 7]
```diff
@@ -715,6 +715,81 @@ preprocessing and query in $\mathcal{O}(1)$ time.
...
+	int n, log2dist;
+	vector<vector<T>> st;
+	/** constructor for the class */
+
+	RMQ(const vector<T> &v) {
+		n = (int)v.size();
+		log2dist = 1 + (int)log2(n);
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:733]: just ceil is fine idk

File: content/4_Gold/Tree_Euler.mdx [hunk 8]
```diff
@@ -715,6 +715,81 @@ preprocessing and query in $\mathcal{O}(1)$ time.
...
+		for (int v : adj[u]) {
+			if (v == p) { continue; }
+			dep[v] = dep[u] + 1;
+			dfs(v, u);
+			et[timer++] = u;
+		}
+	}
+
+  public:
+	/** constructor for the class */
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:772]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
today on "justin is the goat"

File: content/4_Gold/Tree_Euler.mdx (hunk 6 above)
Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:726]: oh i normally use class b/c its shorter to type, but typename makes more sense in this case

Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:726]: oops, will fix

File: content/4_Gold/Tree_Euler.mdx (hunk 7 above)
Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:733]: nah, i ran it on yosupo and setting log2dist to (int) ceil(log2(n)) causes RTE

Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx:735]: LOL WHAT

Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:735]: wait no i’m stupid i think i submitted my lca code to static rmq LMFAO

Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:735]: i’ll change it later 💀

Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:735]: https://judge.yosupo.jp/submission/211411 i lied lmao my filter settings were wrong, it does in fact rte

Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx:735]: rte?

Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:735]: yeah, idk why it does

File: content/4_Gold/Tree_Euler.mdx [hunk 9]
```diff
@@ -715,6 +715,82 @@ preprocessing and query in $\mathcal{O}(1)$ time.
...
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+template <typename T> struct RMQ {
+	int n, log2dist;
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:730]: make class, rename to `SparseTable`, and make vars private

Comment (TheGamingMousse) [reply content/4_Gold/Tree_Euler.mdx:730]: aight bet, gimme a sec

**Review by SansPapyrus683 — APPROVED** [review]

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**